packages/babel-core/src/config/full.js

```diff
@@ -278,17 +278,32 @@ const instantiatePlugin = makeWeakCache(
   },
 );
 
+const validatePreset = (
+  preset: PresetInstance,
+  context: ConfigContext,
+  descriptor: UnloadedDescriptor,
+): void => {
+  if (!context.filename) {
+    const { options } = preset;
+    if (options.test || options.include || options.exclude) {
+      throw new Error(
+        `Preset ${descriptor.name ||
+          ""} requires filename, but it was not passed.`.replace(/\s{2}/, " "),
```

Comment:
Might be nice to link directly to the option:

> Preset ${name} requires a filename be set. See https://babeljs.io/docs/en/options#filename.

Or even:

> Preset ${name} requires a filename be set. For example, if you are calling Babel directly:
>
> ```
> babel.transform(code, { filename: 'file.js', presets: [${name || '/* your preset */'}] });
> ```
>
> See https://babeljs.io/docs/en/options#filename for more information.

(And, as you mentioned, improve this later with user- and preset-specific messages)

Reply:
That delivers way better dev UX. Addressed in ffc6a72.